zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1109 @@
+    };
+
+    use super::{RangeOrdering, SpanningTree};
+
+    #[test]
+    fn range_ordering() {
+        use super::RangeOrdering::*;
+        // Equal
+        assert_eq!(RangeOrdering::cmp(&(0..1), &(0..1)), Equal);
+
+        // Disjoint or contiguous
+        assert_eq!(RangeOrdering::cmp(&(0..1), &(1..2)), LeftFirstDisjoint);
+        assert_eq!(RangeOrdering::cmp(&(1..2), &(0..1)), RightFirstDisjoint);
+        assert_eq!(RangeOrdering::cmp(&(0..1), &(2..3)), LeftFirstDisjoint);
+        assert_eq!(RangeOrdering::cmp(&(2..3), &(0..1)), RightFirstDisjoint);
+
+        // Contained
+        assert_eq!(RangeOrdering::cmp(&(1..2), &(0..3)), LeftContained);
+        assert_eq!(RangeOrdering::cmp(&(0..3), &(1..2)), RightContained);
+        assert_eq!(RangeOrdering::cmp(&(0..1), &(0..3)), LeftContained);
+        assert_eq!(RangeOrdering::cmp(&(0..3), &(0..1)), RightContained);
+        assert_eq!(RangeOrdering::cmp(&(2..3), &(0..3)), LeftContained);
+        assert_eq!(RangeOrdering::cmp(&(0..3), &(2..3)), RightContained);
+
+        // Overlap
+        assert_eq!(RangeOrdering::cmp(&(0..2), &(1..3)), LeftFirstOverlap);
+        assert_eq!(RangeOrdering::cmp(&(1..3), &(0..2)), RightFirstOverlap);
+    }
+
+    fn scan_range(range: Range<u32>, priority: ScanPriority) -> ScanRange {
+        ScanRange::from_parts(
+            BlockHeight::from(range.start)..BlockHeight::from(range.end),
+            priority,
+        )
+    }
+
+    fn spanning_tree(to_insert: &[(Range<u32>, ScanPriority)]) -> Option<SpanningTree> {
+        to_insert.iter().fold(None, |acc, (range, priority)| {
+            let scan_range = scan_range(range.clone(), *priority);
+            match acc {
+                None => Some(SpanningTree::Leaf(scan_range)),
+                Some(t) => Some(t.insert(scan_range)),
+            }
+        })
+    }
+
+    #[test]
+    fn spanning_tree_insert_adjacent() {
+        use ScanPriority::*;
+
+        let t = spanning_tree(&[
+            (0..3, Historic),
+            (3..6, Scanned),
+            (6..8, ChainTip),
+            (8..10, ChainTip),
+        ])
+        .unwrap();
+
+        assert_eq!(
+            t.into_vec(),
+            vec![
+                scan_range(0..3, Historic),
+                scan_range(3..6, Scanned),
+                scan_range(6..10, ChainTip),
+            ]
+        );
+    }
+
+    #[test]
+    fn spanning_tree_insert_overlaps() {
+        use ScanPriority::*;
+
+        let t = spanning_tree(&[
+            (0..3, Historic),
+            (2..5, Scanned),
+            (6..8, ChainTip),
+            (7..10, Scanned),
+        ])
+        .unwrap();
+
+        assert_eq!(
+            t.into_vec(),
+            vec![
+                scan_range(0..2, Historic),
+                scan_range(2..5, Scanned),
+                scan_range(5..6, Historic),
+                scan_range(6..7, ChainTip),
+                scan_range(7..10, Scanned),
+            ]
+        );
+    }
+
```

Comment:
```suggestion

    #[test]
    fn spanning_tree_insert_rfd_span() {
        use ScanPriority::*;

        // This sequence of insertions causes a RightFirstDisjoint on the last insertion,
        // which originally had a bug that caused the parent's span to only cover its left
        // child. The bug was otherwise unobservable as the insertion logic was able to
        // heal this specific kind of bug.
        let t = spanning_tree(&[
            // 6..8
            (6..8, Scanned),
            //       6..12
            // 6..8        8..12
            //         8..10  10..12
            (10..12, ChainTip),
            //          3..12
            //    3..8        8..12
            // 3..6  6..8  8..10  10..12
            (3..6, Historic),
        ])
        .unwrap();

        assert_eq!(t.span(), (3.into())..(12.into()));
        assert_eq!(
            t.into_vec(),
            vec![
                scan_range(3..6, Historic),
                scan_range(6..8, Scanned),
                scan_range(8..10, Historic),
                scan_range(10..12, ChainTip),
            ]
        );
    }

```